tracing-core/src/field.rs

```diff
@@ -572,6 +572,11 @@ impl Field {
     pub fn name(&self) -> &'static str {
         self.fields.names[self.i]
     }
+
+    #[doc(hidden)]
+    pub const fn new_unchecked(i: usize, fields: FieldSet) -> Self {
+        Field { i, fields }
+    }
```

Comment:
> I did figure out that we could write a const fn while loop over the field names array using `[T]::split_first`

Actually, we don't need 1.56 for that, as we could just reimplement that ourselves at MSRV of 1.46. (:

There are still other problems though. The `FieldSet::field` takes its argument as `Q: Borrow<str>`, and it needs to copy the `Identifier` to crate a `Field`. Both of those will generate the following error:

```
error[E0658]: trait bounds other than `Sized` on const fn parameters are unstable
   --> tracing-core/src/field.rs:684:26
    |
684 |     pub const fn field<Q: ?Sized>(&self, name: &Q) -> Option<Field>
    |                        ^
    |
    = note: see issue #57563 <https://github.com/rust-lang/rust/issues/57563> for more information
    = help: add `#![feature(const_fn_trait_bound)]` to the crate attributes to enable
```

We could just add an extra API which fetches the `Field` by `&str` (that seems kinda redundant as a public API though, since `field` already exists). As for the problem of the `Identifier` not being copyable in a `const` context we'd have to get rid of the `&'static dyn Callsite` inside of it, which I think should be doable by casting that pointer to another type before assigning it.

So, I've managed to do this on MSRV 1.46. I've pushed it as a [separate branch here](https://github.com/koute/tracing/commit/6c29c855e4a5edf31ffc16a3040be605b2ee5a15) (untested, and still needs a minor cleanup), so the question is - do you actually want it? I believe this is the best we can do at this point, and that *still* needs an extra API to be added to `tracing-core` (the `field_by_str`) as there's no way to make `field` into a `const fn` since it takes `name` as `&impl Borrow<str>`. (Personally I'd make it non-generic, but that'd be a breaking change.)